### ISSUE TITLE
Incorrect null reference warning if null-check is in another case-guard

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#42057]
**Version Used**: master

**Steps to Reproduce**:
```cs
using System;
#nullable enable
public class C {
    public object? Property;
    public void M(object c) {
        switch(c)
        {
            case C a when a.Property is null:
                break;
            case C a:
                a.Property.ToString(); // warns (none expected)
                break;
        }
    }
}
```

**Expected Behavior**: no warning on ToString() line

**Actual Behavior**: warning CS8602: Dereference of a possibly null reference.

**Comment by jcouv** [comment]
Tagging @gafter to confirm, but I believe this is expected based on language design. The analysis done on `when` clauses doesn't carry over to any other branch.

**Comment by alrz** [comment]
That's true of other analysis as well (like def-assignment), but I was surprised that we don't recognize that, because DAG impl is specifically designed to merge identical `case` labels and only branch on `when` clauses for lowering.

**Comment by gafter** [comment]
We sometimes base the analysis on the DAG, and sometimes on a sequential reading of the cases. If you want something included, better to be explicit about it:

```csharp
case C { Property: null } a:
    break;
case C { Property: { } } a:
    a.Property.ToString();
    break;
```

**Comment by jcouv** [comment]
I'll go ahead and close this isue as by-design.
